src/views/PositionPdfPlot/PositionPdfPlotWidget.tsx

```diff
@@ -70,103 +71,98 @@ const PositionPdfPlotWidget: FunctionComponent<Props> = ({fetchSegment, startTim
         if (visibleTimeEndSeconds === undefined) return undefined
         const i1 = Math.max(0, Math.floor((visibleTimeStartSeconds - startTimeSec) * samplingFrequency))
-        const i2 = Math.min(dataModel.numTimepoints, Math.ceil((visibleTimeEndSeconds - startTimeSec) * samplingFrequency))
+        const i2 = Math.min(numTimepoints, Math.ceil((visibleTimeEndSeconds - startTimeSec) * samplingFrequency))
         return linearPositions.slice(i1, i2)
-    }, [dataModel.numTimepoints, linearPositions, samplingFrequency, startTimeSec, visibleTimeStartSeconds, visibleTimeEndSeconds])
-
-    const margins = useTimeseriesMargins(timeseriesLayoutOpts)
-
-    const panelCount = 1
-    const toolbarWidth = timeseriesLayoutOpts?.hideToolbar ? 0 : DefaultToolbarWidth
-    const { panelWidth, panelHeight } = usePanelDimensions(width - toolbarWidth, height, panelCount, panelSpacing, margins)
-    const timeToPixelMatrix = use1dScalingMatrix(panelWidth, visibleTimeStartSeconds, visibleTimeEndSeconds, margins.left)
-
-    const pixelTimes = useMemo(() => {
-        return convert1dDataSeries([t1, t2], timeToPixelMatrix)
-    }, [t1, t2, timeToPixelMatrix])
-
+    }, [numTimepoints, linearPositions, samplingFrequency, startTimeSec, visibleTimeStartSeconds, visibleTimeEndSeconds])
+    
     const {minValue, maxValue} = useMemo(() => {
         if (!visibleValues) return {minValue: 0, maxValue: 0}
         return {
             minValue: min(visibleValues.map(a => (min(a)))),
             maxValue: max(visibleValues.map(a => (max(a)))),
         }
     }, [visibleValues])
-
+    
     const imageData = useMemo(() => {
         if (!visibleValues) return undefined
         if (minValue === undefined) return undefined
         if (maxValue === undefined) return undefined
-        const N1 = visibleValues.length
-        const N2 = visibleValues[0].length
-        const imageData = new ImageData(N1, N2)
-        let i = 0
-        for (let i2=0; i2<N2; i2++) {
-            for (let i1=0; i1<N1; i1++) {
-                const vv = visibleValues[i1][N2 - 1 - i2]
-                if (vv !== undefined) {
-                    const v = (vv - minValue) / (maxValue - minValue)
-                    const col = colorForValue(v)
-                    imageData.data[i + 0] = col[0]
-                    imageData.data[i + 1] = col[1]
-                    imageData.data[i + 2] = col[2]
-                    imageData.data[i + 3] = 255
-                }
-                else {
-                    imageData.data[i + 0] = 0
-                    imageData.data[i + 1] = 0
-                    imageData.data[i + 2] = 0
-                    imageData.data[i + 3] = 0
-                }
-                i += 4
+        const totalTimePoints = visibleValues.length
+        const totalLinearPositionBuckets = visibleValues[0].length
+        const data = []
+        const colorForValue = getColorForValueFn(minValue, maxValue)
```

Comment:
Pushed the `if undefined` logic into this function, which makes the below code read a little more clearly.